File: leetcode/binary_search/644_Maximum_Average_Subarray_II/codes/bs.py
from typing import List

class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        def is_valid(mid):
            prefix = [0]
            for num in nums:
                prefix.append(prefix[-1] + num - mid)
            
            min_prev = 0
            for i in range(k, len(prefix)):
                if prefix[i] - min_prev >= 0:
                    return True
                min_prev = min(min_prev, prefix[i - k + 1])
            return False

        left, right = min(nums), max(nums)
        epsilon = 1e-5

        while right - left > epsilon:
            mid = (left + right) / 2
            if is_valid(mid):
                left = mid
            else:
                right = mid

        return left

File: leetcode/binary_search/644_Maximum_Average_Subarray_II/codes/bs.c
#include <stdbool.h>
#include <math.h>
#include <stdio.h>

bool isValid(int* nums, int numsSize, int k, double mid) {
    double sum = 0;
    for (int i = 0; i < k; i++) {
        sum += nums[i] - mid;
    }
    if (sum >= 0) return true;

    double prevSum = 0, minPrevSum = 0;
    for (int i = k; i < numsSize; i++) {
        sum += nums[i] - mid;
        prevSum += nums[i - k] - mid;
        minPrevSum = fmin(minPrevSum, prevSum);
        if (sum - minPrevSum >= 0)
            return true;
    }
    return false;
}

double findMaxAverage(int* nums, int numsSize, int k) {
    double left = -10000, right = 10000, eps = 1e-5;
    while (right - left > eps) {
        double mid = (left + right) / 2;
        if (isValid(nums, numsSize, k, mid))
            left = mid;
        else
            right = mid;
    }
    return left;
}